File: scala/src/main/scala/year2020/Day02.scala
package com.ochronus.year2020

import com.ochronus.utils.File

class InputLine(var low: Int, var high: Int, var char: Char, var password: String) {
  def part1: Boolean = {
    val count = password.count(_ == char)
    count >= low && count <= high
  }

  def part2: Boolean = {
    (password(low - 1)) == char ^ (password(high - 1) == char)
  }

}

class Day02 {
  private val lineRegexp = """(\d+)-(\d+) (\w): (\w+)""".r

  private val lines = File.readFileLines("day02.input.txt").map {
      case lineRegexp(low, high, char, password) => new InputLine(low.toInt, high.toInt, char.head, password)
  }

  def part1(): Int = lines.count(_.part1)

  def part2(): Int = lines.count(_.part2)
}

File: python3/day02/day02.py
import re


def parse_line(line):
    m = re.search("(\d+)-(\d+) (\w+): (\w+)", line)
    low = int(m.group(1))
    high = int(m.group(2))
    char = m.group(3)
    pw = m.group(4)
    return (low, high, char, pw)


def part1():
    counter = 0
    for line in open("input.txt").read().split("\n"):
        low, high, char, pw = parse_line(line)
        cnt = pw.count(char)
        if cnt <= high and cnt >= low:
            counter += 1
    return counter


def part2():
    counter = 0
    for line in open("input.txt").read().split("\n"):
        low, high, char, pw = parse_line(line)
        if (pw[low - 1] == char) ^ (pw[high - 1] == char):
            counter += 1
    return counter


print(part1())
print(part2())
